statshandler.go

```diff
@@ -52,12 +78,101 @@ func (h *statshandler) HandleRPC(ctx context.Context, s stats.RPCStats) {
 	case *stats.InPayload:
 		h.jctx.stats.totalInPayloadLength += uint64(s.(*stats.InPayload).Length)
 		h.jctx.stats.totalInPayloadWireLength += uint64(s.(*stats.InPayload).WireLength)
+		if *csvStats {
+			switch v := (s.(*stats.InPayload).Payload).(type) {
+			case *na_pb.OpenConfigData:
+				updateStats(h.jctx, v, false)
+				for idx, kv := range v.Kv {
+					updateStatsKV(h.jctx, false, 0)
+					switch kvvalue := kv.Value.(type) {
+					case *na_pb.KeyValue_UintValue:
+						if kv.Key == "__timestamp__" {
+							var re_c_ts uint64 = 0
+							var re_p_get_ts uint64 = 0
+							if len(v.Kv) > idx+2 {
+								nextKV := v.Kv[idx+1]
+								if nextKV.Key == "__junos_re_stream_creation_timestamp__" {
+									re_c_ts = nextKV.GetUintValue()
+								}
+								nextnextKV := v.Kv[idx+2]
+								if nextnextKV.Key == "__junos_re_payload_get_timestamp__" {
+									re_p_get_ts = nextnextKV.GetUintValue()
+								}
+							}
+
+							//"sensor-path", "sequence-number", "component-id", "sub-component-id", "packet-size", "p-ts", "e-ts", "re-stream-creation-ts", "re-payload-get-ts"))
+							h.jctx.config.CsvStatsJtimon.logger.Printf(
+								fmt.Sprintf("%s,%d,%d,%d,%d,%d,%d,%d,%d\n",
+									v.Path, v.SequenceNumber, v.ComponentId, v.SubComponentId, s.(*stats.InPayload).Length, v.Timestamp, kvvalue.UintValue, re_c_ts, re_p_get_ts))
+						}
+					}
+				}
+			case *gnmi_pb.SubscribeResponse:
+				stat := getKPIStats(v)
+				if stat != nil && stat.Timestamp != 0 {
+					path := stat.SensorName + ":" + stat.Streamed_path + ":" + stat.Path + ":" + stat.Component
+					h.jctx.config.CsvStatsJtimon.logger.Printf(
+						fmt.Sprintf("%s,%d,%d,%d,%d,%d,%d,%d,%d\n",
+							path, stat.SequenceNumber, stat.ComponentId, stat.SubComponentId,
+							s.(*stats.InPayload).Length, stat.notif_timestamp, int64(stat.Timestamp*uint64(1000000)),
+							int64(stat.re_stream_creation_timestamp*uint64(1000000)),
+							int64(stat.re_payload_get_timestamp*uint64(1000000)),
+						),
+					)
+				}
+			}
+		}
 	case *stats.InTrailer:
 	case *stats.End:
 	default:
 	}
 }
 
+func getKPIStats(subResponse *gnmi_pb.SubscribeResponse) *kpiStats {
+
+	stats := new(kpiStats)
+	notfn := subResponse.GetUpdate()
+	if notfn == nil {
+		return nil
+	}
+	stats.notif_timestamp = notfn.Timestamp
+	extns := subResponse.GetExtension()
+
+	if extns != nil {
+		extn := extns[0]
```

Comment:
Can we search by extension ID ? If Juniper adds other extensions, it need not always add them as an ordered list.

Reply:
What should be the extension ID ??

Reply:
done

Reply:
I am still not seeing the logic to fetch by extension ID. Please check and resolve later, for now it is fine as we have only one extension.